### DISCUSSION TITLE
Stuck on "Waiting for application startup" when adding long-running loop in startup event. Can Background Tasks be used in startup?

### DISCUSSION THREAD

**Discussion by davidtgq** [fastapi/fastapi#6526] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
import asyncio
import websockets
import json
from fastapi import FastAPI, WebSocket, WebSocketDisconnect

app = FastAPI()

subscribers = []

@app.websocket("/ws")
async def websocket_endpoint(ws: WebSocket):
	await ws.accept()
	subscribers.append(ws)
	while True:
		try:
			await ws.receive_text()
		except WebSocketDisconnect:
			print(f"Client disconnected: {ws.client}")
			return


@app.on_event("startup")
async def startup():
	await asyncio.create_task(mirror_ticks())


async def mirror_ticks():
	async with websockets.connect("wss://ws-feed.pro.coinbase.com") as ws:
		await ws.send(json.dumps({
			"type": "subscribe",
			"product_ids": ["BTC-USD", "ETH-USD"],
			"channels": ["ticker"]
		}))
		while True:
			data = await ws.recv()
			for sub in subscribers:
				sub.send_text(data)
```

### Description

I want to augment an existing script to also mirror the incoming data, and am trying to add FastAPI/websockets to implement this. However, I don't know the correct syntax/mechanism to make it work well with my original script.

I tried this: https://github.com/tiangolo/fastapi/issues/543#issuecomment-547087693 But it just results in getting stuck at "Waiting for application startup."

I guess I need something like `asyncio.gather(fast_api(), mirror_ticks())`, or something like Background Tasks https://fastapi.tiangolo.com/tutorial/background-tasks/  in the startup?

### Operating System

Linux, Windows

### FastAPI Version

0.68

### Python Version

3.9

### Additional Context

I'm currently running it with `uvicorn {this_filename}:app` (so there's currently no `.run()`)
The `mirror_ticks()` that it's stuck in is working, since if you do `print(data)` inside the loop, it prints what I expect it to.

**Comment by wotori** [comment]
I met the same problem ("Waiting for application startup") when forgot to connect to VPN with the data base.

**Comment by YuriiMotov** [comment, marked as answer]
> ```
> @app.on_event("startup")
> async def startup():
> 	await asyncio.create_task(mirror_ticks())
> ```
You shouldn't **await** it, but just save it in `app.state` and cancel this task on app shutdown:

```python
@app.on_event("startup")
async def startup():
    app.state.mirror_task = asyncio.create_task(mirror_ticks())

@app.on_event("shutdown")
async def shutdown():
    app.state.mirror_task.cancel()
    try:
        await app.state.mirror_task
    except asyncio.CancelledError:
        pass
```

or with lifespan (preferred):

```python
@asynccontextmanager
async def lifespan(app: FastAPI):
    # Startup logic
    task = asyncio.create_task(mirror_ticks())
    app.state.mirror_task = task

    try:
        yield
    finally:
        # Shutdown logic
        task.cancel()
        try:
            await task
        except asyncio.CancelledError:
            pass

app = FastAPI(lifespan=lifespan)
```

PS: didn't check the code, but it should work
